### ISSUE TITLE
[BUG] Computing ESP Using Density Matrix Expressed w.r.t. MOs

### ISSUE THREAD

**Issue by maximilianvz** [theochem/gbasis#143]
**Describe the bug**

As I was making the tutorial notebook that I've proposed in PR #142, I believe I came across a bug that comes about when trying to compute the electrostatic potential for a system using the density matrix expressed with respect to molecular orbitals. I passed a density matrix expressed w.r.t. molecular orbitals of shape (14, 14), whereas the density matrix expressed w.r.t. atomic orbitals would've been of shape (18, 18). Even though I provided a `transform` argument (a (14, 18) matrix) to tell this higher level function that I want to work in terms of the MO basis, it threw an error saying that my one-electron density matrix was the wrong shape (it was expected to be (18, 18)). I believe the problem is that the presence or absence of the `transform` matrix when calling `electrostatic_potential` is irrelevant when it comes to checking the shape of the one-electron density matrix.

I don't think there was a test in `gbasis/tests` to check this behaviour, despite it being apparently possible to carry out this operation in the [notes](https://github.com/theochem/gbasis/blob/95f7cec39b74c7f260e7823c3af9cac43ca3d1ac/docs/notes.pdf).

**To Reproduce**

*see screenshot

**Expected behaviour**

I would expect no exception of this sort to be raised when the density matrix expressed w.r.t. molecular orbitals is passed to `electrostatic_potential` along with a proper `transform` argument. This can likely be (at least partially) fixed by modifying the conditional statements [1. here](https://github.com/theochem/gbasis/blob/95f7cec39b74c7f260e7823c3af9cac43ca3d1ac/gbasis/evals/electrostatic_potential.py#L103), [2. here](https://github.com/theochem/gbasis/blob/95f7cec39b74c7f260e7823c3af9cac43ca3d1ac/gbasis/evals/electrostatic_potential.py#L109), and [3. here](https://github.com/theochem/gbasis/blob/95f7cec39b74c7f260e7823c3af9cac43ca3d1ac/gbasis/evals/electrostatic_potential.py#L124) to include an `and transform is None`. If `transform is None`, then the shape of the one-electron density matrix should be (# AOs, # AOs). However, when `transform` is passed, I believe it would have to be of shape (# MOs, # MOs). [This is confirmed in the documentation for `electrostatic_potential.py`](https://github.com/theochem/gbasis/blob/95f7cec39b74c7f260e7823c3af9cac43ca3d1ac/gbasis/evals/electrostatic_potential.py#L24C9-L25C67). Right now, it doesn't look like it matters if `transform` is passed or not; the one-electron density matrix is always expected to be of shape (# AOs, # AOs) according to the `sum(cont.num_sph * cont.num_seg_cont for cont in basis) != one_density_matrix.shape[0]` clause.

**Screenshots**

<img width="539" alt="Screenshot 2023-12-24 at 4 46 31 PM" src="https://github.com/theochem/gbasis/assets/83951518/84da9cfd-d7da-444a-9533-b6cff225a6c4">

**System Information:**
 - OS:
 - Python version:
 - NumPy version:
 - SciPy version:

**Additional context**

**Comment by leila-pujal** [comment]
Me and @maximilianvz looked at this. All the high order functions in gbasis have available the transform argument where you can pass a transformation matrix to go from one basis to another (e.g AO to MO). If you pass this argument it is assumed your one-electron density matrix would be in the transformed basis. Focusing on computing the electrostatic potential, this referers to equation 49 in [david's notes](https://github.com/theochem/gbasis/blob/master/docs/notes.pdf), if I am not wrong eq 53 in Paul's notes, and this part of the code 

https://github.com/theochem/gbasis/blob/c8268710b77c706f02487e907339f904c12397c9/gbasis/evals/electrostatic_potential.py#L133-L137 

So line 133 calls the point_charge integral. If you pass the transformation matrix it should give you back the electrostatic potential in the molecular orbital basis set. The error that Max is getting is because there is an assertion to check if the one-electron density matrix has the same number of dimensions as the number of Atomic orbitals. In the case of using a transform, this shouldn't be checked because that one electron density matrix is expressed in the basis of molecular orbitals which can have a different shape. It looks like there is not a check for using electrostatic_potential function passing a transformation (https://github.com/theochem/gbasis/blob/master/tests/test_electrostatic_potential.py). Taking into account eq 49 from David's notes: 
$` - (- \sum_A \frac{Z_A}{|\mathbf{R}_C - \mathbf{R}_A|} + \sum_{ab} \gamma_{ab} \int \phi_a(\mathbf{r}) \frac{-1}{|\mathbf{r} - \mathbf{R}_C|} \phi_b(\mathbf{r}) d\mathbf{r} )  =   \sum_A \frac{Z_A}{|\mathbf{R}_C - \mathbf{R}_A|} - \sum_{ab} \gamma_{ab} \int \phi_a(\mathbf{r}) \frac{1}{|\mathbf{r} - \mathbf{R}_C|} \phi_b(\mathbf{r}) d\mathbf{r} `$

The transform matrix is used when this is computed:
$` \int \phi_a(\mathbf{r}) \frac{1}{|\mathbf{r} - \mathbf{R}_C|} \phi_b(\mathbf{r}) d\mathbf{r} `$

And the error comes from doing the sum $`  \sum_{ab} \gamma_{ab} `$. My question would be, is $` \gamma_{ab} `$ allowed to be in a molecular orbitals basis set? I have the feeling that because of the form of the point charge integral it may not be possible. If this is the case then we should delete the transform argument for this function. If $` \gamma_{ab} `$  could be the one electron density matrix in the base of molecular orbitals, then we should delete the assertation. @PaulWAyers and @FarnazH, if you could give us any insights into this then I think we would be able to solve this issue.

**Comment by PaulWAyers** [comment]
First, the output is on points, so there isn't such a thing as the "electrostatic potential in the molecular orbital basis" or not. The electrostatic potential is evaluated at points.

What I think is clear (sorry to be picky about words) is the question is whether one can compute the electrostatic potential using the density-matrix in either the AO or the MO basis. I would say yes. The MO basis should *usually* be the same dimension as the AO basis, unless one has truncated the MO basis. (E.g., maybe your MOs are actually natural orbitals, and you only care about matrix elements for occupied MOs.)  

I will need to look more at the example, but it's very surprising to me that there are 14 MOs and 18 AOs. That is not normal, and a test---at least a warning---associated with that case is something I'm in favor of.

**Comment by PaulWAyers** [comment]
Looking at the notebook there is something very strange. Your MOs are not even orthogonal or normalized, as they are just a random transformation. So I don't think the evaluations are meaningful.

If you don't want to have to read in a calculation, use an ANO basis:  in those cases, the orbitals *are* natural orbitals, and the AOs and MOs are the same. There are (obviously) the same number of both. You could make a unitary rotation of the basis if you wanted to show a transformation.

I'm coming to this issue late, and I'm curious what @FarnazH thinks, but I'd prefer to use a real example (imported using `IOData`). I think we were using formaldehyde, $\ce{CH2O}$ in `grid`.

**Comment by maximilianvz** [comment]
@PaulWAyers, that's a very good catch. I was expecting some feedback on the order of upping my examples, so anything you can suggest would be much appreciated. 

The only reason I used this Kr example was because it seems a decent number of tests in the repo set things up as I did in the notebook. For example, in [`test_evaluate_density`](https://github.com/theochem/gbasis/blob/c8268710b77c706f02487e907339f904c12397c9/tests/test_density.py#L62), I believe the same setup as I erroneously implemented in the notebook was used; the `transform` matrix is of shape (14,18) such that 14 MOs are generated from 18 AOs. I recognize that it in general, the MO basis is usually the same dimension as the AO basis, but when I saw this example in the tests, I thought maybe I was missing something, so I just threw it in the notebook as a draft. In the case that the MO basis *is* the same dimension as the AO basis, this "bug" that I'm reporting wouldn't be a thing anymore. I agree that we could implement a warning when the MO basis is truncated. If I'm correct in my interpretation of the tests, maybe those need to be revamped, but you'd be a better judge than myself.

**Comment by PaulWAyers** [comment]
No problem, I spoke too soon...I saw Kr and thought it was a "real" example, then read more and realized it was just used as a template....I typed before I thought. 

I think it is very safe to assume the MO and AO basis have the same dimension. When they do not, the transformation isn't a transformation...it's a projection (onto a smaller space). I updated my comment above to reflect this. I think the more substantive question, which pertains to the tests as well as the example, is whether we should use a *real* example. In the era where `gbasis` was being written `iodata` was (at best) half-baked, but now we can use `iodata` to generate some data files for tests (without having an explicit `iodata` dependence in the tests).
